### ISSUE TITLE
Implement external plugin system.

### ISSUE THREAD

**Issue by dsw88** [byu-oit/handel#302]
People would like Handel services for things like MongoDB and RabbitMQ, which AWS doesn't provide native services for.

Handel should focus on the core AWS services, but it should offer a plugin model that allows people to define service provisioners that can be imported into the deployment lifecycle.

**Comment by ThatJoeMoore** [comment]
I've been thinking about how to implement this.

The code samples given below may conflict with each other - I'm basically walking through my thought/design process, so each is an evolution on the previous in order to solve the new problem.

**tl;dr:** While a super-basic plugins implementation would be easy to add, there are some complicated things we should at least think about.

# The plugins used in a handel stack must be explicitly declared

This helps make the handelfile more understandable, as one can see at a glance which plugins are being used.  It also solves a host of problems related to service names.

## What it looks like

Have a `plugins` section in `handel.yml`:

```yaml
version: 1
plugins:
- foobar-handel-plugin
```

## Alternatives

We could implement things Yeoman-style and require plugins to have a certain pattern to their names, as well as declaring a certain keyword in their package.json.  So, for example, a `foobar` plugin would have a (minimal) package.json like:

```json
{
  "name": "handel-foobar-plugin",
  "version": "1.0.0",
  "keywords": ["handel-plugin"]
}
```

This has the downside of making Handel more dependent on the user's local configuration, and making it harder to know which plugins must be installed for a given Handelfile to run.  It might, in fact, make it very difficult for something like Handel-Codepipeline to work properly with plugins, as it would have to somehow discover which plugins are being used.

# We should have the user namespace the plugins they use

If we force the user to define a namespace for each plugin they use, we automatically avoid service name conflicts. By making the user define the namespace, we can make some things more convenient for both them and us. They'll always have a simple name for each plugin, and we don't have to worry about plugins declaring conflicting 'short' namespaces (I'll explain this below).

## What it looks like

`handel.yml`
```yaml
version: 1

plugins:
  foobar: foobar-handel-plugin

environments:
  dev:
    foobar:
      type: foobar::foo
      foo: 'bar'
```

## Alternatives

There are two alternatives I can think of: using the NPM package name as the namespace and having each plugin declare a namespace for itself.

If we use the package name, we end up with one of two problems: either we have really long namespaces (`acme-company-foobar-handel-plugin`) or we have to constrain what package names can look like (ie, if the package name ends with `-handel-plugin`, we omit it from the namespace), which I feel makes things less obvious - You could have a `acme-company-foobar-handel-plugin` package and a `acme-company-foobar` on NPM - how do I know which to install?

If we allow the plugin to declare a 'short' namespace - like `acme-company-foobar-handel-plugin` could use `acme-foobar`, we run into the possibility of having two plugins conflict with their short names.  We would then have to do something like allowing the user to use either the 'short' namespace or the 'long' namespace, and forcing them to use the 'long' namespace in cases of conflict.  This could make our code very complicated, and cause Handelfiles to suddenly become invalid by simply installing a conflicting plugin or having a plugin update add a new 'short' namespace that conflicts with an existing one.

# Plugin installation and version management.

One real gotcha here is plugin versions - How do we make sure that, if a plugin issues a new version, we don't break everything?  I don't really want to have to have users in non-node contexts have a package.json listing all of the plugins they're using, but I feel like we should have some way of making sure that the version being pulled in matches the one we expect (or, is within a range of expected plugins). We could have them put a version or version range in their Handel file, and then have Handel be in charge of downloading the proper plugins for a project using the NPM apis. This, however, introduces some complexity on our side.  It'd be much easier for us to force the user to have every plugin they want to use installed as a global dependency, but then it becomes really easy for things to break, and what if they have multiple projects that need different versions of a given plugin?

I'm leaning towards putting Handel in charge of installing downloading and installing plugins. It adds complexity, and may require the user to know more about the version of the plugin they're using, but we might be able to add some tooling around that.

## What it looks like

We could have many different ways of declaring the plugin version in `handel.yml`:

### Inline/@
```yaml
plugins:
  foo: foobar-handel-plugin@^1.2.1
```
This is nice and concise, but maybe hard to read.

We could even have the option of not declaring an explicit version at all, and just using the latest version, unless the user specifies one.

```yaml
plugins:
  foo: foobar-handel-plugin
  bar: bar-handel-plugin@^1.2.1
```

### Version Key
```yaml
plugins:
  foo: 
    name: foobar-handel-plugin
    version: ^1.2.1
```

### Useability Concerns

Compared to just using `npm install`, it becomes kind of a pain for the user to have to go out and find the version of a plugin, then define an appropriate version range if they want to avoid having to keep it up to date manually. 

I'm thinking we could implement an NPM-style Handel command:

`handel add-plugin foobar-handel-plugin`, which would insert the appropriate name and version number into `handel.yml`. This would use the NPM programmatic APIs, or use the `npm` cli itself.

We clould also have a `handel install-plugins` to download all plugins without actually running a Handel command.

### Downloading

We could download the files into a `handel_plugins` directory in the project root and lay it out like `node_modules`.  Then, the Handel runtime looks there to resolve its plugins. When the user runs a CLI command, we check and see if we have all of our plugins downloaded, download any missing ones, then go on our merry way.

## Alternatives

We could...not solve it? That doesn't sound like a great strategy long-term, though.

Perhaps we have a separately-versioned file, kind of like `package-lock.json` or `yarn.lock` that, if present, enforces version ranges for Handel plugins?

I don't really love any of these solutions, but they're the best I can come up with right now.

# Plugins should tell Handel what services they provide

We should let Plugins organize themselves howerver they wish, and provide a programmatic API for declaring what services they support.

## What it looks like

We would use the "main" script declared in the plugin's package.json:

**Somewhere inside Handel**
```js
function registerPlugin(name) {
  let plugin = require(`handel_plugins/${name}`);//This might not work, we may need to load and read package.json ourselves 

  return plugin.initializeHandelPlugin();
}
```

`initializeHandelPlugin` (name subject to change) would return an object mapping plugin names to implementations, or perhaps, for lazy-loading, a function which returns the implementation.

**Plugin `package.json`**
```json
{
  "main": "lib/handel-plugin.js"
}
```

**Plugin `lib/handel-plugin.js`**
```js

const fooService = require('./services/foo');
const barService = require('./services/bar');
const bazService = require('./services/baz');

exports.initializeHandelPlugin = function() {
  return {
    foo: fooService,
    bar: barService,
    baz: bazService
  }
}
```

# Plugin Spec Versioning

However Handel loads plugins, we may need to provide a way to say which 'version' of the plugin spec a given plugin implements.  Otherwise, we essentially lock ourselves into one service contract forever, unable to add, subtract, or modify.  We could either have this version be declared in a field in the package.json, which we look at, or we can have the declaration come in the Javascript that bootstraps the plugin, like `initHandelPluginV1`, `initHandelPluginV2`, etc. That's not necessarily bad, though, since we could then offer better type completion for each version.

## What it looks like

**Plugin `lib/handel-plugin.js`**
```js

const fooService = require('./services/foo');

exports.initHandelPluginV1 = function() {
  return {
    foo: fooService
  }
}
```

# We could eventually separate out the 'standard' Handel plugin

I'm not sure we'd want to do this, but it's something to think about - we could have a `handel-core` package, with the main Handel lifecycle logic and interfaces, a `handel-stdlib`, with the main services, and a `handel` package (or rename it to `handel-cli`), which contains the CLI and depends on `handel-stdlib`.  This would make everything independently versionable, but I also don't know how much that really gives us.

**Comment by ThatJoeMoore** [comment]
So, I've been thinking about starting to implement this, and I decided to take a look at how VSCode handles extensions, and I really like how they're doing it.

They've got a series of extensions they've added to `package.json`. VSCode then uses those to determine what each extension package is capable of, then it lazily instantiates the parts of the package that it needs.

They've got their extensions documented at https://code.visualstudio.com/docs/extensionAPI/extension-manifest.  Some highlights:

* They use the standard 'engines' section to specify compatibility with VS Code versions.
* They add a 'contributes' section to list the different things the extension can do
* They add an 'activationEvents' section to list the events that should cause VS Code to initialize the extension
* The module specified in the standard 'main' field must, at a minimum, expose an 'activate' function which accepts a VS Code 'ExtensionContext' object. The function uses that object to actually register its various behaviors.

I kind of like this model. It'd allow us to easily see what each extension/plugin (I'm starting to lean towards 'extension' as the proper term) is capable of doing without having to initialize it. I think that that is a Good Thing, as it allows us to easily do things like checking that we have implementations for every service we see in the Handel environment.

On the other hand, I see some benefit to keeping all of the main definitions in code - single source of truth and whatnot.  One thing I definitely think we should do is enforce a rule that you must specify 'handel' in your engines section, to help us deal with future versioning issues easily.

So, since I've now talked myself back into favoring a code-based approach, here's what I'm thinking things would look like:

**Project `handel.yml`**
```yaml
handel.yml

version: 1

extensions:
  foobar: foobar-handel-extension@^1.2.1

environments:
  dev:
    foobar:
      type: foobar::foo
      foo: 'bar'
```

**Extension `package.json`**
```json
{
  "name": "foobar-handel-extension",
  "version": "1.2.3",
  "main": "dist/extension.js",
  "engines": {
    "handel": "^1.0.0"
  },
  "dependencies": {
    "handel-extensions-api": "^1.0.0",
    ... others ...
  }
}
```

**Extension `extension.ts`**
```typescript
import { ExtensionContext, HookType } from 'handel-extensions-api';
import { customValidator } from './custom-validator';
import { customPostDeployer } from './post-deploy';
import { customService } from './custom-service';

export function activateHandelExtension(context: ExtensionContext) {
  // Register a validator
  context.hook(HookType.VALIDATION, customValidator);
  // Register some sort of post-deploy cleanup or something
  context.hook(HookType.POST_DEPLOY, customPostDeployer);
  // Register a service
  context.service('custom', customService);
}
```

An alternative 'hook' implementation might look like this:

```typescript
  context.hooks.validation(customValidator);
  context.hooks.postDeploy(customPostDeployer);
```

I'm honestly leaning towards this one, as it would allow us to have a different argument type for each type of hook. Yay good type-safety and IDE suggestion support!

I kind of like where all this is going.  Any thoughts from anyone, especially @dsw88 ?

**Comment by dsw88** [comment]
We had an in-person discussion about the last issue. We think we'll just not worry about implementing the full extension style that VS Code uses, since we're ok with initializing the declared plugins during application startup.

We agreed that some sort of handel-extensions-api module is required for plugins written in TS. This module would export a large part of what is currently in the datatypes file in Handel.

We also talked about removing the hooks from the proposed extension spec, since we think the current use-case for those should be implemented in a different way outside of Handel. We can always add the concept of hooks back in if required.

@ThatJoeMoore did I miss anything?

**Comment by ThatJoeMoore** [comment]
@dsw88 Nope.

Here's my todo list. Each one will have its own PR so that we have plenty of time to look at each piece individually.

- [x] Add Lerna support, so that we can have multiple packages in the same repo (#371).
- [ ] Create the extensions-api package, containing public type definitions.
- [ ] Write a simplistic extension.
- [ ] Add actual support for extensions behind a hidden CLI flag to allow testing and initial plugin development.
- [ ] Turn extensions on by default.

**Comment by dsw88** [comment]
Well we never added a feature flag and rather just left extensions on for anyone who knew how to use them.

I added documentation for extensions: http://handel.readthedocs.io/en/latest/extensions/using-extensions.html. I also added documentation for writing extensions: http://handel.readthedocs.io/en/latest/extensions/writing-extensions.html.

I think there are a few things left before we close this issue:

- [ ] Add whitelist to account config file for trusted extensions
- [ ] Add log somewhere that extensions can use.
